code_challenges/code_wars/move_zeros_to_the_end.py

```diff
@@ -0,0 +1,9 @@
+# https://www.codewars.com/kata/52597aa56021e91c93000cb0
+
+
+def move_zeros_to_end(numbers):
+    for number in numbers:
+        if number == 0:
+            numbers.remove(number)
+            numbers.append(number)
+    return numbers
```

Comment:
This is a much simpler solution, but has some faults too.

`for number in numbers` is a loop over numbers, and .remove and .append are both changing the length of numbers.  You should  avoid changing the length of a collection within a loop over it.  In this case it is working because the start and end length between iterations is remaining the same.

Another thing that might not be obvious is that .remove is also a loop, so is iterating over the list again to find the 0.  So the time complexity of this solution is 0(n²).  Meaning the operations worst case are the square of the input length. This isn't all bad because the space complexity of this solution is only n since the size in memory isn't increasing with input length.  It depends on what you're going for.  Most case one shoots for less than or equal to O(n) for both time and space.

In python .remove has to look at every value sequentially and check if it is equal to what you are asking for it to remove, and then once it finds it, removes it.  Removing it is also a time complex operation, because one by one ever value has to be shifted to fix the index.  So best case the number of operations are the length of the list.

Now consider a couple solutions:
```python
def move_zeros_to_end(numbers):
    result_zeros = []
    result_numbers = []
    for number in numbers:
        if number:
            result_numbers.append(number)
        else:
            result_zeros.append(number)
    return result_numbers + result_zeros
```
^ This one is O(n) in time complexity, and O(n) in space complexity, since both the operations and the memory usage increase linearly with input length. 
```python
def move_zeros_to_end(numbers):
    i, j, stop = 0, 1, len(numbers) - 1
    while i < stop:
        if numbers[i] == 0:
            while j < stop and numbers[j] == 0:
                j += 1
            numbers[i], numbers[j] = numbers[j], numbers[i]
        i, j = i + 1, i + 1 if i > j else j
    return numbers
```
^ Is called a Two pointer (also known as 'Sliding Window') solution.  Like your original, it's `n` in space complexity, everything is moved in place, but this one is O(n) in time complexity.

Some reading
https://www.freecodecamp.org/news/big-o-notation-why-it-matters-and-why-it-doesnt-1674cfa8a23c/
https://medium.com/swlh/two-pointer-technique-solving-array-problems-at-light-speed-56a77ee83d16